django_advanced_queries/covid_19/models.py

```diff
@@ -2,12 +2,285 @@
 from __future__ import unicode_literals
 
 from django.db import models
+from django.db.models.functions import Coalesce
+
+
+class SickPersonsMixin(object):
+    """Mixin that responsible fetching sick objects from manager.
+
+    Help us fetching sick persons from each person that is patient model.
+    """
+
+    def get_sick_records(self, patient_id_attribute):
+        """Get sick objects from current class.
+
+        patient_id_attribute (str): Reference to patient id of current model.
+
+        Example:
+            self.get_sick_objects("person__patients_details")
+            "person__patients_details" - Points to the current model
+            patient id.
+        """
+        # Extract latest exam.
+        latest_exam = \
+            MedicalExaminationResult.objects.get_patient_examination_results(
+                patient_id=models.OuterRef(patient_id_attribute)
+            )[:1]
+
+        # Attach for each record - the latest exam result he got.
+        records_with_latest_results = self.annotate(
+            latest_result=models.Subquery(latest_exam),
+        )
+
+        # Get only results that are not healthy/dead.
+        sick_records = records_with_latest_results.filter(
+            ~models.Q(latest_result=MedicalExaminationResult.RESULT_HEALTHY) &
+            ~models.Q(latest_result=MedicalExaminationResult.RESULT_DEAD)
+        )
+
+        return sick_records
+
+
+# Managers
+
+
+class PatientManager(models.Manager, SickPersonsMixin):
+    """Custom Queryset methods to patient model."""
+
+    def filter_by_examinations_results_options(self, results):
+        # Get the patients that have result which appear in results list.
+        return self.filter(
+            medical_examination_results__result__in=results
+        ).distinct()
+
+    def get_highest_num_of_patient_medical_examinations(self):
+        # For each patient - attach it's exam count.
+        # It possible to count medical_examination_results because it queryset.
+        exams_king = self.annotate(
+            exam_count=models.Count("medical_examination_results")
+        ).order_by("-exam_count").values("exam_count").first()
+
+        return exams_king["exam_count"]
+
+    def get_sick_patients(self):
+        return self.get_sick_records(patient_id_attribute="id")
+
+    def filter_by_examined_hospital_workers(self, hospital_workers):
+        sick_workers_patients = self.filter(
+            medical_examination_results__examined_by__in=hospital_workers
+        ).values("id")
+
+        return self.filter(id__in=sick_workers_patients)
```

Comment:
why do you filter again using the ids from the previous query instead of returning the result of previous query?